File: leetcode-cs/0900_0999/leetcode-no984.cs
// problem: https://leetcode.cn/problems/string-without-aaa-or-bbb/?envType=problem-list-v2&envId=3rBpGyas 

using System.Text;

namespace leetcodeNo984;

using System;

public class Solution2 {
    // kind version
    public string StrWithout3a3b(int a, int b) {
        string res = "";
        char[] cand = ['a', 'b'];
        int[] cnt = [a, b];
        int last = a > b ? 1 : 0;
        while (cnt[0] > 0 || cnt[1] > 0) {
            var cur = last ^ 1;
            if (cnt[last] >= cnt[cur]) {
                res += cand[cur];
                cnt[cur]--;
            } else {
                res += cand[cur];
                if (cnt[cur] > 1) {
                    res += cand[cur];
                }

                cnt[cur] -= 2;
            }

            last = cur;
        }

        return res;
    }
}

public class Solution {
    // evil version
    public string StrWithout3a3b(int a, int b) {
        var sb = new StringBuilder(a + b);
        int[] cnt = [a, b];
        var cur = a > b ? 0 : 1;
        do {
            sb.Append((char)('a' + cur));
            if (cnt[cur ^ 1] < cnt[cur] && cnt[cur]-- > 1)
                sb.Append((char)('a' + cur));
        } while (--cnt[(cur ^= 1) ^ 1] > 0 || cnt[cur] > 0);

        return sb.ToString();
    }
}

public class Test {
    public static void Run() {
        var s = new Solution();
        Console.WriteLine(s.StrWithout3a3b(1, 2));
        Console.WriteLine(s.StrWithout3a3b(4, 1));
    }
}

//
// Created By jing At 11/17/25 20:11:14
//

File: leetcode-cpp/0900_0999/leetcode-no984.cpp
// problem: https://leetcode.cn/problems/string-without-aaa-or-bbb/
#include <iostream>
#include <ranges>
#include <algorithm>
#include <vector>
#include "../utils/vector_helper.h"


using namespace std;

class Solution {
public:
    string strWithout3a3b(int a, int b) {
        string res;
        res.reserve(a + b);
        int cnt[] = {a, b};
        int cur = a > b ? 0 : 1;
        do {
            res += static_cast<char>('a' + cur);
            if (cnt[cur ^ 1] < cnt[cur] && cnt[cur]-- > 1)
                res += static_cast<char>('a' + cur);
        } while (--cnt[(cur ^= 1) ^ 1] > 0 || cnt[cur] > 0);
        return res;
    }
};

int main(int argc, char *argv[]) {
    Solution test;
    cout << test.strWithout3a3b(1, 2) << endl;
    cout << test.strWithout3a3b(4, 1) << endl;
}

//
// Created By ASUS At 2025-11-18 01:01
//
